Don't over-allocate memory for DTLS fragments

julek-wolfssl opened the pull request:
Don't reserve the full message length of memory. Instead we only allocate memory for the fragments that we have already received. We also dynamically combine memory fragments when we receive overlap.

julek-wolfssl committed:
Don't over-allocate memory for DTLS fragments

Don't reserve the full message length of memory. Instead we only allocate memory for the fragments that we have already received. We also dynamically combine memory fragments when we receive overlap.

rizlik commented on src/internal.c:
```diff
@@ -8176,132 +8176,271 @@ void DtlsTxMsgListClean(WOLFSSL* ssl)
-                                     msg->msg, &bytesLeft, heap);
-            if (newFrag == NULL)
-                return MEMORY_E;
+    /* Expand the larger bucket if data bridges the gap between cur and next */
```
asaiu `if data *does no* bridges...`, otherwise always expand next

julek-wolfssl replied:
No. If it bridges the gap, then expand the larger of `cur` and `next`. Otherwise, we always expand `cur`.

rizlik commented on src/internal.c:
```diff
@@ -8176,132 +8176,271 @@ void DtlsTxMsgListClean(WOLFSSL* ssl)
-            msg->fragList = newFrag;
+    if (combineNext) {
+        /* Move data already in bucket first. Use memmove since newBucket may be
+         * next. */
```
afaiu newBucket is always next if combineNext is true

julek-wolfssl replied:
If `combineNext` is true `newBucket` can still be `cur` if `cur->sz >= next->sz` is true. I want to increase the larger bucket to minimize memory used while combining the data.

rizlik commented on src/internal.c:
```diff
@@ -8176,132 +8176,271 @@ void DtlsTxMsgListClean(WOLFSSL* ssl)
-            return 0;
+    if (newOffset == offset) {
+        /* data comes first */
+        if (newOffsetEnd <= offsetEnd) {
```
nit: can this be < ?

julek-wolfssl replied:
It can be if `data` extends into `next`. Copying from `next` modifies `newOffsetEnd`.

rizlik reviewed (changes requested):
Overall the PR looks good to me. I suggest extensive testing as fragmentation code is always tricky.
Do you think it may be a good moment to move all dtls fragmentation logic to dtls.c?

rizlik commented on src/internal.c:
```diff
@@ -8176,132 +8176,271 @@ void DtlsTxMsgListClean(WOLFSSL* ssl)
-/* Create a DTLS Fragment from *begin - end, adjust new *begin and bytesLeft */
-static DtlsFrag* CreateFragment(word32* begin, word32 end, const byte* data,
-                                byte* buf, word32* bytesLeft, void* heap)
+static DtlsFragBucket* CreateFragBucket(word32 offset, const byte* data,
```
Consider using a common prefix for all DtlsMsg functions

julek-wolfssl replied:
Changing to use the common prefix.

julek-wolfssl commented on src/internal.c:
```diff
@@ -8176,132 +8176,271 @@ void DtlsTxMsgListClean(WOLFSSL* ssl)
+    else {
+        /* cur -> data */
+        word32 curOffsetEnd = cur->offset + cur->sz;
+        if (newBucket != cur) {
```
That is why I also set `*chosenBucket` :)

julek-wolfssl commented:
>  Do you think it may be a good moment to move all dtls fragmentation logic to dtls.c?

That is a good idea. I might do that next.

julek-wolfssl committed:
Use a union and struct for padding in DtlsFragBucket

Zero length arrays are not allowed so `byte padding[0]` is not a valid member. Changed to use a union and struct instead.

julek-wolfssl committed:
Expose API for src/api.c

dgarske commented on wolfssl/internal.h:
```diff
@@ -4419,24 +4424,36 @@ typedef struct DtlsRecordLayerHeader {
     word32          sz;        /* Length of whole message      */
     byte            type;
+    byte            fragBucketListCount;
+    byte            ready:1;
```
Windows is unhappy `c:\jenkins\workspace\prb-windows-test-v2\wolfssl\internal.h(4455): warning C4200: nonstandard extension used : zero-sized array in struct/union [C:\jenkins\workspace\PRB-windows-test-v2\wolfssl.vcxproj]`

julek-wolfssl replied:
Fixed by using a union+struct instead of a padding array.

julek-wolfssl replied:
Actually needed to add a `pragma` to ignore this warning.

julek-wolfssl committed:
Disable warning for MSVC

julek-wolfssl commented on src/dtls13.c:
```diff
@@ -677,9 +677,14 @@ static int Dtls13DetectDisruption(WOLFSSL* ssl, word32 fragOffset)
-    if (ssl->dtls_rx_msg_list == NULL ||
-        ssl->dtls_rx_msg_list->fragSz != fragOffset) {
-        return 1;
+    if (ssl->dtls_rx_msg_list == NULL) {
```
Fixed.

dgarske commented on wolfssl/internal.h:
```diff
@@ -4434,24 +4439,40 @@ typedef struct DtlsRecordLayerHeader {
+         * DtlsFragBucket buffer. */
+        byte padding[DTLS_HANDSHAKE_HEADER_SZ];
+    } m;
+/* Ignore "nonstandard extension used : zero-sized array in struct/union" 
```
Trailing whitespace:
```[check-source-text] [2 of 18] [9c6698b3b3]
8 bit, trailing spaces or tabs:
./wolfssl/internal.h:4455:/* Ignore "nonstandard extension used : zero-sized array in struct/union"
weird control chars, hard tabs, CRs, trailing whitespace:
./wolfssl/internal.h:4455:/* Ignore "nonstandard extension used : zero-sized array in struct/union"```

dgarske commented on wolfssl/internal.h:
```diff
@@ -4434,24 +4439,40 @@ typedef struct DtlsRecordLayerHeader {
+#ifdef _MSC_VER
+#pragma warning(disable: 4200)
+#endif
+    byte buf[];
```
Why not use `byte *buf` here? Instead of "unsized array"?

dgarske reviewed (changes requested):
Thanks, looks good otherwise.

dgarske reviewed (approved):
Looks good. @rizlik any feedback?

rizlik reviewed (approved):
nice job!

dgarske commented on src/internal.c:
```diff
@@ -8187,132 +8187,281 @@ void DtlsTxMsgListClean(WOLFSSL* ssl)
-            if (newFrag == NULL)
-                return MEMORY_E;
+    {
+        DtlsFragBucket* tmp = (DtlsFragBucket*)XREALLOC(*chosenBucket,
```
@julek-wolfssl this adds a new XREALLOC requirement for DTLS on embedded that was not there previously. Not all platforms have XREALLOC.

```
../Middlewares/Third_Party/wolfSSL_wolfSSL_wolfSSL/wolfssl/src/internal.c:8324:48: warning: implicit declaration of function 'XREALLOC'; did you mean 'XMALLOC'? [-Wimplicit-function-declaration]
 8324 |         DtlsFragBucket* tmp = (DtlsFragBucket*)XREALLOC(*chosenBucket,
```

Perhaps we could do one of the following:
1) Setup a generic fallback XREALLOC use case to malloc, copy and free old?
2) Refactor this to use XMALLOC/XFREE only if XREALLOC is not defined?

julek-wolfssl replied:
I like solution number 1 more since that will make the code cleaner. I made a note to take care of this.
